src/lib.rs

```diff
@@ -142,80 +135,112 @@ struct CoreSDK<WP> {
-                let task_token = work.task_token.clone();
-                debug!(
-                    task_token = %fmt_task_token(&task_token),
-                    "Received workflow task from server"
-                );
-
-                let (next_activation, run_id) = self.instantiate_or_update_workflow(work)?;
-
-                if next_activation.more_activations_needed {
-                    self.pending_activations.push(PendingActivation {
-                        run_id,
-                        task_token: task_token.clone(),
+    fn poll_task(&self, req: PollTaskReq) -> Result<Task> {
+        let task_queue = req.task_queue;
+        // The poll needs to be in a loop because we can't guarantee tail call optimization in Rust
+        // (simply) and we really, really need that for long-poll retries.
+        loop {
+            if req.workflows {
+                // We must first check if there are pending workflow tasks for workflows that are
+                // currently replaying, and issue those tasks before bothering the server.
+                if let Some(pa) = self.pending_activations.pop() {
+                    Span::current().record("pending_activation", &format!("{}", &pa).as_str());
+
+                    let next_activation =
+                        self.access_wf_machine(&pa.run_id, move |mgr| mgr.get_next_activation())?;
+                    let task_token = pa.task_token.clone();
+                    if next_activation.more_activations_needed {
+                        self.pending_activations.push(pa);
+                    }
+                    return Ok(Task {
+                        task_token,
+                        variant: next_activation.activation.map(Into::into),
                     });
                 }
+            }
 
-                Ok(Task {
-                    task_token,
-                    variant: next_activation.activation.map(Into::into),
-                })
+            if self.shutdown_requested.load(Ordering::SeqCst) {
+                return Err(CoreError::ShuttingDown);
             }
-            // Drain pending activations in case of shutdown.
-            Err(CoreError::ShuttingDown) => self.poll_task(task_queue),
-            Err(e) => Err(e),
-            Ok(PollTaskResponse::ActivityTask(_)) => Err(CoreError::UnexpectedResult),
-        }
-    }
 
-    #[instrument(skip(self))]
-    fn poll_activity_task(&self, task_queue: &str) -> Result<Task> {
-        if self.shutdown_requested.load(Ordering::SeqCst) {
-            return Err(CoreError::ShuttingDown);
-        }
+            // TODO: There's an interesting question here about whether activities or workflows
+            //  should take priority in the event that the caller is OK with either, and we
+            //  could potentially poll both and just proceed with whichever returns first... but
+            //  that starts getting tricky. Perhaps we should not allow specifying both.
```

Comment:
From lang side we need to be able to request both in order to not spawn multiple threads.
I don't think there should be precedence of either workflows or activities, better to poll on both concurrently.

Reply:
The way it's built right now, we can't poll for both which is problematic.

Reply:
Heh, yes, hence the TODO. Polling for both is problematic as I discovered here. See my other reply